src/core/Akka/Dispatch/ChannelSchedulerExtension.cs

```diff
@@ -0,0 +1,389 @@
+                        newWorkerToStart--;
+                    }
+                }
+            }
+
+            //reschedule
+            if (!_cts.IsCancellationRequested)
+            {
+                var interval = controlWorkerCount > 0 || (reqWorkerCount - newWorkerToStart) > 0
+                    ? _workInterval / _workStep
+                    : _workInterval * _workStep;
+                _timer.Change(interval, Timeout.Infinite);
+            }
+        }
+
+        private void Worker(object state)
+        {
+            DoWork((int)state);
+        }
+
+        private int DoWork(int workerId)
+        {
+            var highCount = 0;
+            var normalCount = 0;
+            var lowCount = 0;
+            var idleCount = 0;
+
+            int c;
+            int rounds = 0;
+            int roundWork;
+            int roundClean = 0;
+
+            //maybe implement max work count and/or a deadline
+
+            _threadPriority = TaskSchedulerPriority.Idle;
+            try
+            {
+                do
+                {
+                    rounds++;
+                    roundWork = 0;
+
+                    c = _highScheduler.ExecuteAll();
+                    highCount += c;
+                    roundWork += c;
+
+                    c = _normalScheduler.ExecuteMany(_maxWork);
+                    normalCount += c;
+                    roundWork += c;
+
+                    c = roundWork > 0
+                        ? _lowScheduler.ExecuteSingle()
+                        : _lowScheduler.ExecuteMany(_maxWork);
+                    lowCount += c;
+                    roundWork += c;
+
+                    //if there was no work then only execute background tasks 
+                    if (c == 0)
+                    {
+                        c = _idleScheduler.ExecuteSingle();
+                        idleCount += c;
+                        roundWork += c;
+                    }
+
+                    roundClean = roundWork == 0 ? roundClean + 1 : 0;
+                }
+                while (roundClean < 2 && !_cts.IsCancellationRequested);
+            }
+            catch
+            {
+                //ignore error
+            }
+            finally
+            {
+                _threadPriority = TaskSchedulerPriority.None;
+            }
+
+            //worker stopped
+
+            var total = highCount + normalCount + lowCount + idleCount;
+
+            //todo push to metrics: workerId, total, highCount, normalCount, lowCount, idleCount
+
+            return total;
+        }
+
+        public void Dispose()
+        {
+            _idleScheduler.Dispose();
+            _lowScheduler.Dispose();
+            _normalScheduler.Dispose();
+            _highScheduler.Dispose();
+
+            _cts.Cancel();
+            _timer.Dispose();
+        }
+
+        sealed class PriorityTaskScheduler : TaskScheduler, IDisposable
+        {
+            readonly Channel<Task> _channel;
```

Comment:
Good point, I will take a look the the Channel source code too.

In the end the queue should lock-free queued and dequeued, 
where adding side should have the heavier processing (splitting in priorities)

Reply:
I looked up and yes there is a brief lock. 

The lock will not have much perf. issue on the executing thread itself,
but of course on locking threads. Don't know why it is not implemented in a free-lock style. 
Maybe the lock collation just don't happen that often to take it into effect?
I trust ["stephentoub" ](https://github.com/stephentoub) in his decision to use a lock here.

The `lock` statement is on itself is not "bad" but misused it can be very fast.

The main benefit of the Channel class is following line, it is reducing the idle cpu to near zero:
https://github.com/dotnet/runtime/blob/386f87139c7fe48f6b733e3cae6d09128c827192/src/libraries/System.Threading.Channels/src/System/Threading/Channels/UnboundedChannel.cs#L295

Reply:
I rechecked now the `UnboundedChannel.TryWrite` code
1) it would be absolutely hard not to use a sync lock
2) Collisions are very unlikely, even under heavy queue load